bee-ledger/src/workers/pruning/prune.rs

```diff
@@ -29,172 +31,230 @@ static NUM_PRUNINGS: AtomicUsize = AtomicUsize::new(0);
-
-        metrics.receipts = milestone_data_metrics.receipts;
-
-        // Add unconfirmed data to the delete batch.
-        let batch_unconfirmed_data = Instant::now();
-        let unconfirmed_data_metrics = batch::batch_prunable_unconfirmed_data(storage, &mut batch, index)?;
-        timings.batch_unconfirmed_data = batch_unconfirmed_data.elapsed();
-
-        metrics.messages += unconfirmed_data_metrics.prunable_messages;
-        metrics.edges += unconfirmed_data_metrics.prunable_edges;
-        metrics.indexations += unconfirmed_data_metrics.prunable_indexations;
-
-        // Remove old SEPs from the storage.
-        //
-        // **WARNING**: This operation must come before the batch is committed!
-        //
-        // TODO: consider batching deletes rather than using Truncate. Is one faster than the other? Do we care if its
-        // atomic or not?
-        let truncate_old_seps = Instant::now();
-        Truncate::<SolidEntryPoint, MilestoneIndex>::truncate(storage).expect("truncating solid entry points failed");
-        timings.truncate_curr_seps = truncate_old_seps.elapsed();
-
-        // Execute the batch operation.
-        let batch_commit = Instant::now();
-        storage
-            .batch_commit(batch, true)
-            .map_err(|e| Error::Storage(Box::new(e)))?;
-        timings.batch_commit = batch_commit.elapsed();
-
-        // Update the pruning index.
-        tangle.update_pruning_index(index);
-
-        // Write the updated snapshot info to the storage.
-        let timestamp = SystemTime::now()
-            .duration_since(SystemTime::UNIX_EPOCH)
-            .expect("error creating timestamp")
-            .as_secs();
-        let mut snapshot_info = storage::fetch_snapshot_info(storage)
-            .map_err(|e| Error::Storage(Box::new(e)))?
-            .ok_or(Error::MissingSnapshotInfo)?;
-        snapshot_info.update_pruning_index(index);
-        snapshot_info.update_timestamp(timestamp);
-        storage::insert_snapshot_info(storage, &snapshot_info).map_err(|e| Error::Storage(Box::new(e)))?;
-
-        timings.full_prune = full_prune.elapsed();
-
-        debug!("{:?}.", metrics);
-        debug!("{:?}", confirmed_data_metrics);
-        debug!("{:?}", unconfirmed_data_metrics);
-        debug!("{:?}.", timings);
-        debug!(
-            "Entry point index now at {} with {} solid entry points..",
-            index, num_next_seps
-        );
-        debug!("Pruned milestone {}.", index);
 
-        bus.dispatch(PrunedIndex { index });
+        num_pruned_bytes +=
+            prune_milestone::<_, true>(index, tangle, storage, bus, &mut timings, &mut metrics, config).await?;
+
+        log::debug!("Pruned {num_pruned_bytes}/{num_bytes_to_prune} bytes.");
     }
 
-    if start_index == target_index {
-        info!("Pruned milestone {}.", start_index);
+    Ok(())
+}
+
+/// Prunes a single milestone.
+async fn prune_milestone<S: StorageBackend, const BY_SIZE: bool>(
```

Comment:
I really think we shouldn't do that `const BY_SIZE: bool` thing. We don't gain anything from it as it's far from being a hot path and it's unnecessarily duplicating a lot of code. A regular bool parameter would be better IMO.

Reply:
But it is a hot LOOP actually. Have a look at the `prune_by_size` function. The pruning continues until `num_bytes_to_prune` is reached.

Reply:
I might however not pass it up to `prune_milestone` :+1:  which is in fact duplicating a lot of code.

Reply:
I think this is completely negligible compared to the magnitude of the actual pruning operations tbh

Reply:
So you are concerned about binary size? I feel like that's also negligible but I might be wrong. The question is what we optimize for. Since this is a library I would say if we have a to weigh performance against binary size, I'ld untertake everything to squeeze out maximum performance - even if tiny.

Reply:
Changed